netebpfext/net_ebpf_ext_sock_addr.c

```diff
@@ -871,15 +894,33 @@ _Requires_exclusive_lock_held_(_net_ebpf_ext_sock_addr_lock) static void _net_eb
 static void
 _net_ebpf_ext_purge_lru_contexts(BOOLEAN delete_all)
 {
+    ebpf_epoch_state_t epoch_state = {0};
     KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+    ebpf_epoch_enter(&epoch_state);
+
     _net_ebpf_ext_purge_lru_contexts_under_lock(delete_all);
+    ebpf_hash_table_destroy(_net_ebpf_ext_connect_context_hash_table);
+    _net_ebpf_ext_connect_context_hash_table = NULL;
+
+    ebpf_epoch_exit(&epoch_state);
     ExReleaseSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock, old_irql);
 }
 
-static void
+static ebpf_result_t
 _net_ebpf_ext_insert_connection_context_to_list(_Inout_ net_ebpf_extension_connection_context_t* connection_context)
 {
+    ebpf_result_t result = EBPF_SUCCESS;
+    ebpf_epoch_state_t epoch_state = {0};
     KIRQL old_irql = ExAcquireSpinLockExclusive(&_net_ebpf_ext_sock_addr_lock);
+    ebpf_epoch_enter(&epoch_state);
+
+    // Insert into hash table
+    result = ebpf_hash_table_update(
+        _net_ebpf_ext_connect_context_hash_table,
+        (uint8_t*)connection_context,
+        (uint8_t*)&connection_context,
```

Comment:
any particular reason the value is a pointer to a pointer?

Reply:
I wanted to leave the existing LRU logic to ensure that any potentially stale connection_contexts would always be cleaned up. This required having this sock_addr hook still own the underlying memory of the entry. So I needed to store just a pointer to the entry itself, instead of having the hash table own the underlying memory of the entry.